### ISSUE TITLE
cpyext `Py_Is` implementation does pointer comparison

### ISSUE THREAD

**Issue by gitlab-importer** [pypy/pypy#4044]
In Heptapod by @ngoldbaum on Dec 15, 2023, 17:24

It looks like Py_Is was introduced to pypy 3.10 in a merge with upstream about a year ago. However, the implementation inherited from CPython is incorrect and needs to be fixed for pypy. See https://foss.heptapod.net/pypy/pypy/-/blob/branch/py3.10/pypy/module/cpyext/include/object.h?ref_type=heads#L125

**Comment by mattip** [comment]
@ngoldbaum hi (we moved to github). I am not sure this is a PyPy issue. Is there supposed to be a python guarantee that any float values, say 3.1415, if passed from Python to C, will end up with the same `PyObject` in C? Is `float('nan')` somehow more special?

**Comment by ngoldbaum** [comment]
My opening this issue was based on the comment [here](https://bugs.python.org/issue43753#msg390435) and replies in that thread. The behavior around `is` comparisons not necessarily being the same as pointer comparisons was a main motivation for adding `Py_IS` to the C API in the first place.

> Is there supposed to be a python guarantee that any float values, say 3.1415, if passed from Python to C, will end up with the same PyObject in C? Is float('nan') somehow more special?

No, there's nothing special about floats, and CPython does not make that guarantee for `float(NaN)`:

```
Python 3.11.6 (main, Nov 27 2023, 12:13:52) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.nan is float('nan')
False

```
>>> float('nan') is float('nan')
False
``` 

`pypy` however, does mandate that using `is` with any two `NaN` floats will return `True` (and the same is true of any floats with identical bit representations, from [here](https://doc.pypy.org/en/latest/cpython_differences.html#object-identity-of-primitive-values-is-and-id).

As an author of C code writing against the CPython C API, I would expect `Py_IS` to implement the same semantics as the python-level `is` operator, which the current implementation on pypy inherited from CPython does not do.

**Comment by cfbolz** [comment]
fwiw, I agree with @ngoldbaum that `Py_IS` should behave the same as Python's `is`.

**Comment by cfbolz** [comment]
sorry, just realized that my statement is possible ambiguous :sweat_smile:. I think in PyPy's `Py_IS` should return True for two NaNs, and also for eg two integers with the same value.

**Comment by mattip** [comment]
In PR #4843 I changed the implementation to use python-level `is`.

**Comment by mattip** [comment]
@ngoldbaum do you need this backported to Python3.9?

**Comment by ngoldbaum** [comment]
No, it's not in the 3.9 CPython C API so I don't think it makes sense for pypy to ship it there.

**Comment by mattip** [comment]
Closing, PyPy 7.3.15 was released with #4843. Please reopen or open a new issue if the fix is insufficient.
